File: examples/order_workflow.rs
use std::time::Duration;

use aws_durable_execution_sdk as durable;
use durable::Branch;

/// Processes an order: validates, fans out inventory and payment checks in
/// parallel, waits for a warehouse, then requires manager approval via callback.
async fn handler(
    event: serde_json::Value,
    ctx: durable::DurableContext,
) -> Result<serde_json::Value, durable::BoxError> {
    let validated = ctx
        .step(|_| async { Ok(event.get("orderId").cloned()) })
        .name("validate-order")
        .await?;

    let prepared = ctx
        .parallel(vec![
            Branch::new("check-inventory", |child| async move {
                child
                    .step(|_| async { Ok(true) })
                    .name("inventory")
                    .await
                    .map_err(Into::into)
            }),
            Branch::new("reserve-payment", |child| async move {
                child
                    .step(|_| async { Ok(true) })
                    .name("payment")
                    .await
                    .map_err(Into::into)
            }),
        ])
        .name("prepare-order")
        .max_concurrency(2)
        .await?;

    ctx.wait(Duration::from_secs(30))
        .name("warehouse-processing")
        .await?;

    let approval = ctx
        .wait_for_callback::<String, _, _>(|_cb_ctx, _cb_id| async { Ok(()) })
        .name("manager-approval")
        .await?;

    if validated.is_some() {
        let shipped = ctx
            .step(|_| async { Ok("shipped".to_owned()) })
            .name("ship-order")
            .await?;
        Ok(serde_json::json!({ "shipped": shipped, "branches": prepared.len() }))
    } else {
        Ok(serde_json::json!({ "status": "skipped", "approval": approval }))
    }
}

#[tokio::main]
async fn main() -> Result<(), lambda_runtime::Error> {
    lambda_runtime::tracing::init_default_subscriber();
    durable::run(handler).await
}

File: examples/order_workflow.go
// Command order-workflow demonstrates a durable order processing workflow
// built with the AWS Durable Execution SDK for Go.
//
// It exercises the core primitives naturally:
//   - Step for validation, with the result checkpointed for replay
//   - Parallel for concurrent shipment preparation (label + tracking)
//   - Wait for a cooling-off period between preparation and charging
//   - WaitForCallback for an optional manager approval
//   - An if statement wrapping a durable call for the conditional approval
package main

import (
	"errors"
	"fmt"
	"time"

	"github.com/aws/aws-durable-execution-sdk-go/durable"
)

// OrderEvent is the workflow input.
type OrderEvent struct {
	OrderID         string   `json:"orderId"`
	CustomerID      string   `json:"customerId"`
	Amount          float64  `json:"amount"`
	Items           []string `json:"items"`
	RequireApproval bool     `json:"requireApproval"`
}

// OrderResult is the workflow output.
type OrderResult struct {
	Status   string `json:"status"`
	Label    string `json:"label"`
	Tracking string `json:"tracking"`
}

func handler(ctx durable.Context, event OrderEvent) (OrderResult, error) {
	// Validate the order before doing any work.
	if _, err := durable.Step(ctx, "validate-order", func(sc durable.StepContext) (bool, error) {
		if event.Amount <= 0 {
			return false, errors.New("order amount must be positive")
		}
		return true, nil
	}); err != nil {
		return OrderResult{}, err
	}

	// Prepare the shipment in parallel: the label and tracking number are
	// generated concurrently and looked up by branch name afterwards.
	shipment, err := durable.Parallel(ctx, "prepare-shipment", []durable.Branch[string]{
		{
			Name: "generate-label",
			Func: func(branch durable.Context) (string, error) {
				return durable.Step(branch, "label", func(sc durable.StepContext) (string, error) {
					return fmt.Sprintf("https://labels.example.com/%s", event.OrderID), nil
				})
			},
		},
		{
			Name: "generate-tracking",
			Func: func(branch durable.Context) (string, error) {
				return durable.Step(branch, "tracking", func(sc durable.StepContext) (string, error) {
					return fmt.Sprintf("TRK-%s", event.OrderID), nil
				})
			},
		},
	})
	if err != nil {
		return OrderResult{}, fmt.Errorf("prepare shipment: %w", err)
	}
	label, _ := shipment.Result("generate-label")
	tracking, _ := shipment.Result("generate-tracking")

	// Cooling-off period before charging the customer.
	if err := durable.Wait(ctx, "cooling-off", 5*time.Second); err != nil {
		return OrderResult{}, fmt.Errorf("wait interrupted: %w", err)
	}

	// Pause for a human approval only when the order requires it.
	if event.RequireApproval {
		if _, err := durable.WaitForCallback[string](ctx, "manager-approval", func(sc durable.StepContext, callbackID string) error {
			// TODO: notify the approver with callbackID
			return nil
		}, durable.WithCallbackTimeout(24*time.Hour)); err != nil {
			return OrderResult{}, fmt.Errorf("approval failed: %w", err)
		}
	}

	return OrderResult{
		Status:   "FULFILLED",
		Label:    label,
		Tracking: tracking,
	}, nil
}

func main() {
	durable.Start(handler)
}
